### PR TITLE
abci: remove lock protecting calls to the application interface

### LINKED ISSUE #7073: abci: remove global mutex around local client
## Problem Definition

This was brought up in a similar [issue](https://github.com/tendermint/tendermint/issues/6899). Currently, Tendermint uses a mutex around ABCI calls so that they are serialized. This was done out of safety and simplicity. However, this restricts the ability for apps to implement any form of concurrency between the applications. It should thus really be the responsibility of application to handle mutexes.

## Proposal

Remove the mutex here:
https://github.com/tendermint/tendermint/blob/e53f92ba9cfceb49c1588775f6a9c700c22d4c0a/abci/client/local_client.go#L18

We should make sure that the change in responsibility is well documented

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by tychoish**
Closes #7073

As part of the 0.36 cycle we've discussed and decided to remove the mutex in tendermint that protects the ABCI application. First, applications should be able to be responsible for their own concurrency control, and can make more fine-grained decisions about concurrent use than tendermint ever could. Second, I've observed in recent weeks as we've been making this change that the mutex wasn't applied particularly consistently in many cases (e.g. multiple "local" connections to the application had multiple locks, etc.) so this will give more consistent experiences across ABCI execution environments, and simplifies the tendermint ABCI handling code.

### CODE REVIEW

File: abci/server/socket_server.go [hunk 1]
```diff
@@ -67,33 +69,40 @@ func (s *SocketServer) OnStop() {
...
 
 // deletes conn even if close errs
 func (s *SocketServer) rmConn(connID int) error {
 	s.connsMtx.Lock()
 	defer s.connsMtx.Unlock()
+	return s.unsafeRmConn(connID)
+}
+
+func (s *SocketServer) unsafeRmConn(connID int) error {
+	if closer, ok := s.connsClose[connID]; ok {
```
Comment (creachadair) [inline abci/server/socket_server.go:98]: Given that we're already under the lock here, and `net.Conn` implements `io.Closer`, do we need this separate map? It appears if we find a connection in `s.conns` we're already dropping it and closing it.

I see various closer functions plumbed around as parameters, but I maybe have missed where they're originating. Are these "close a related thing" someplace?

Comment (tychoish) [reply abci/server/socket_server.go:98]: `unsafeRmConn` does three things:
- cleans up the map(s)
- call the cancel function for the context that the connection handling threads are running for each connection
- closes the actual connection . 

This is more confusing than it needs to be and I think I can make something work that's more simple.

**Comment by thanethomson** [comment]
This closes #7073 right?

File: UPGRADING.md [hunk 2]
```diff
@@ -2,6 +2,22 @@
...
 
+## v0.36
+
+### ABCI Changes
+
+#### ABCI++
+
+#### ABCI Mutex
+
+In previous versions of ABCI, Tendermint was prevented from making
```
[the remark this answers is not in the document]
Comment (tychoish) [reply UPGRADING.md:13]: I believe that this is really an application concern in terms of making sure that the application doesn't introduce some racey/non-deterministic behavior. 

I think the problems we've seen in the past are "tendermint making concurrent calls to the app exercised application bugs".

[the remark this answers is not in the document]
Comment (marbar3778) [reply UPGRADING.md:13]: so an app can set it up in any way they want? might make sense to document this. Building an app docs need updating as this is something to be careful with .

[the remark this answers is not in the document]
Comment (tychoish) [reply UPGRADING.md:13]: That's the point of this note. I've added some more text, but let me know if we should change it more.

**Review by creachadair — APPROVED** [review]

**Comment by tychoish** [comment]
> This closes #7073 right?

Yes. In combination with https://github.com/tendermint/tendermint/commit/61a81279bd6ebfded58979783629b2916195fca3, tendermint will allow concurrent calls to ABCI applications, and defer concurrency control to the application.

**Pull request merged**